test/e2e/upgrades/serviceaccount_admission_controller_migration.go

```diff
@@ -0,0 +1,135 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	e2eauth "k8s.io/kubernetes/test/e2e/auth"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+const (
+	podBeforeMigrationName = "pod-before-migration"
+	podAfterMigrationName  = "pod-after-migration"
+)
+
+// ServiceAccountAdmissionControllerMigrationTest test that a pod is functioning before and after
+// a cluster upgrade.
+type ServiceAccountAdmissionControllerMigrationTest struct {
+	pod *v1.Pod
+}
+
+// Name returns the tracking name of the test.
+func (ServiceAccountAdmissionControllerMigrationTest) Name() string {
+	return "[sig-auth] serviceaccount-admission-controller-migration"
+}
+
+// Setup creates pod-before-migration which has legacy service account token.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Setup(f *framework.Framework) {
+	t.pod = createPod(f, podBeforeMigrationName)
+	inClusterClientMustWork(f, t.pod)
+}
+
+// Test waits for the upgrade to complete, and then verifies pod-before-migration
+// and pod-after-migration are able to make requests using in cluster config.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	ginkgo.By("Waiting for upgrade to finish")
+	<-done
+
+	ginkgo.By("Starting post-upgrade check")
+	podBeforeMigration, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.TODO(), podBeforeMigrationName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	if podBeforeMigration.GetUID() != t.pod.GetUID() {
+		framework.Failf("Pod %q GetUID() = %q, want %q.", podBeforeMigration.Name, podBeforeMigration.GetUID(), t.pod.GetUID())
+	}
+	if podBeforeMigration.Status.ContainerStatuses[0].RestartCount != 0 {
+		framework.Failf("Pod %q RestartCount = %d, want 0.", podBeforeMigration.Name, podBeforeMigration.Status.ContainerStatuses[0].RestartCount)
+	}
+	if len(podBeforeMigration.Spec.Volumes) != 1 || podBeforeMigration.Spec.Volumes[0].Secret == nil {
+		framework.Failf("Pod %q Volumes[0].Secret = nil, want non-nil.", podBeforeMigration.Name)
+	}
+
+	ginkgo.By("Checking pod-after-migration makes successful requests using in cluster config")
+	podAfterMigration := createPod(f, podAfterMigrationName)
+	if len(podAfterMigration.Spec.Volumes) != 1 || podAfterMigration.Spec.Volumes[0].Projected == nil {
+		framework.Failf("Pod %q Volumes[0].Projected.Sources = nil, want non-nil.", podAfterMigration.Name)
+	}
+	inClusterClientMustWork(f, podAfterMigration)
+
+	ginkgo.By("Finishing post-upgrade check")
+}
+
+// Teardown cleans up any remaining resources.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Teardown(f *framework.Framework) {
+	// rely on the namespace deletion to clean up everything
+}
+
+func inClusterClientMustWork(f *framework.Framework, pod *v1.Pod) {
+	var logs string
+	if err := wait.Poll(1*time.Minute, 20*time.Minute, func() (done bool, err error) {
```

Comment:
this will always wait 1 minute before checking the first time... should this use PollImmediate and/or a shorter interval?

Reply:
since i was not checking whether we have seen tokens in log, PollImmediate might only execute once if there is no matched logs.

i changed to PollImmediate and checking that we at least see one token in the log.